hydra_provider/src/input/serial_input.rs

```diff
@@ -26,29 +32,83 @@ impl SerialInput {
                 .port_name
                 .clone()
         };
+        let config = format!("serial:{}:{}", port, baud_rate);
+        let link = mavlink::connect::<uorocketry::MavMessage>(config.as_str())?;
 
-        info!("Using serial port '{port}'");
+        Ok(SerialInput { link })
+    }
+}
 
-        let port = serialport::new(&port, baud_rate)
-            .timeout(Duration::new(30, 0))
-            .open()
-            .with_context(|| format!("Failed to open serial connection '{port}'"))?;
+impl SerialInput {
+    pub fn read_write_loop(&self, send: Sender<InputData>) -> ! {
+        thread::scope(|s| {
```

Comment:
If anyone is confused on how I'm able to send and receive from two separate threads without angering the borrow checker, let me know, and I can explain a bit more what is happening here.